internal/catalog/internal/controllers/workloadhealth/controller.go

```diff
@@ -0,0 +1,214 @@
+package workloadhealth
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/hashicorp/consul/internal/catalog/internal/controllers/nodehealth"
+	"github.com/hashicorp/consul/internal/catalog/internal/types"
+	"github.com/hashicorp/consul/internal/controller"
+	"github.com/hashicorp/consul/internal/resource"
+	pbcatalog "github.com/hashicorp/consul/proto-public/pbcatalog/v1alpha1"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+var (
+	errNodeUnreconciled            = errors.New("Node health has not been reconciled yet")
+	errNodeHealthInvalid           = errors.New("Node health has invalid reason")
+	errNodeHealthConditionNotFound = fmt.Errorf("Node health status is missing the %s condition", nodehealth.StatusConditionHealthy)
+)
+
+// The NodeMapper interface is used to provide an implementation around being able to
+// map a watch event for a Node resource and translate it to reconciliation requests
+// for all Workloads assigned to that node.
+type NodeMapper interface {
+	// MapNodeToWorkloads will take a Node resource and return controller requests
+	// for all Workloads associated with the Node.
+	MapNodeToWorkloads(ctx context.Context, rt controller.Runtime, res *pbresource.Resource) ([]controller.Request, error)
+
+	// TrackWorkload instructs the NodeMapper to associate the given workload
+	// ID with the given node ID.
+	TrackWorkload(workloadID *pbresource.ID, nodeID *pbresource.ID)
+
+	// RemoveWorkloadTracking instructs the Nodemapper to forget about any
+	// association it was tracking for this workload.
+	RemoveWorkloadTracking(workloadID *pbresource.ID)
```

Comment:
naming nit: `UntrackWorkload` shorter/simpler/obviously pairs with `TrackWorkload`?

Reply:
Sounds good to me. If you happen to see the service endpoints PR (still in draft) I will make similar updates to the workload selection tracker there too.